impute_constant/impute_constant.tex

```diff
@@ -0,0 +1,181 @@
+    showstringspaces=false,
+    showtabs=false,                  
+    tabsize=2
+}
+
+\lstset{style=mystyle}
+
+\newcommand{\silvia}[1]{{ {\color{blue}{(silvia)~#1}}}}
+\newcommand{\grace}[1]{{ {\color{purple}{(grace)~#1}}}}
+
+\newcommand{\MultiSet}{\mathrm{MultiSet}}
+\newcommand{\len}{\mathrm{len}}
+\newcommand{\din}{\texttt{d\_in}}
+\newcommand{\dout}{\texttt{d\_out}}
+\newcommand{\T}{\texttt{T} }
+\newcommand{\F}{\texttt{F} }
+\newcommand{\Relation}{\texttt{Relation}}
+\newcommand{\X}{\mathcal{X}}
+\newcommand{\Y}{\mathcal{Y}}
+\newcommand{\True}{\texttt{True}}
+\newcommand{\False}{\texttt{False}}
+\newcommand{\clamp}{\texttt{clamp}}
+\newcommand{\function}{\texttt{function}}
+\newcommand{\float}{\texttt{float }}
+\newcommand{\questionc}[1]{\textcolor{red}{\textbf{Question:} #1}}
+
+
+\title{Privacy Proofs for OpenDP: Impute Constant Transformation}
+\author{Grace Tian}
+\date{Summer 2021}
+\begin{document}
+
+
+\maketitle
+\tableofcontents
+
+\section{Algorithm Implementation}
+\subsection{Code in Rust}
+The current OpenDP library contains the \texttt{make\_impute\_constant} function implementing the impute constant function. This is defined in lines 62-75 of the file \texttt{impute.rs} in the Git repository (\url{https://github.com/opendp/opendp/blob/21-impute/rust/opendp/src/trans/impute.rs#L62-L75}).
+
+\includegraphics[width=\textwidth]{make_impute_constant.jpg}
+
+
+\subsection{Pseudo Code in Python}\label{sec:pseudocode}
+
+\subsubsection*{Preconditions}
+To ensure the correctness of the output, we require the following preconditions:
+
+\begin{itemize}
+    \item \textbf{User-specified types:}
+    \begin{itemize}
+        \item Variable \texttt{\texttt{constant}} must be of type \texttt{DA::NonNull}
+        \item Type \texttt{DA} must have traits \texttt{ImputableDomain}. % Unless the null method is defined, this means that \texttt{DA} is either \texttt{OptionNullDomain<AllDomain<T>>} or \texttt{InherentNullDomain<AllDomain<T>>}.
+        \item \texttt{DA::NonNull} must have traits \texttt{Clone}
+    \end{itemize}
+\end{itemize}
+
+\subsubsection*{Postconditions}
+
+\begin{itemize}
+    \item Either a valid \texttt{Transformation} is returned or an error is returned.
+\end{itemize}
+
+
+\begin{lstlisting}[language=Python,  escapechar=|]
+def make_impute_constant(constant : DA::NonNull):
+    # instead of VectorDomain(DA), we add ::new() to get a new instance of DA. This is bcause DA has the ImputableDomain trait. Discuss among interns.
+    input_domain = VectorDomain(DA::new())
+    output_domain = VectorDomain(AllDomain(DA::NonNull))
+    input_metric = SymmetricDistance()
+    output_metric = SymmetricDistance()
+    
+    # check constant for nullity first
+    assert(not constant.is_null); # not DA::is_null(constant) |\label{line:null}|
+    def Relation(d_in: u32, d_out: u32) -> bool: |\label{line:rel}|
+        return d_out >= d_in*1
+
+    def function(data: Vec[DA::Carrier]) -> Vec[DA::NonNull]: |\label{line:fn}|
```

Comment:
1. is this type signature correct?

Reply:
This looks mechanically correct. We have DA, the Atomic Domain, the domain inside VectorDomain.
- the input data is a vector of DA's associated "Carrier" type
- the output data is a vector of DA's associated "NonNull" type; aka the type after imputation